File: proxy-android/proxy/src/main/java/com/zeticai/melangelm/model/ChatModels.kt
package com.zeticai.melangelm.model

/**
 * OpenAI-compatible chat request/response data models.
 * Compatible with the /v1/chat/completions endpoint.
 */

data class ChatMessage(
    val role: String,       // "system" | "user" | "assistant"
    val content: String
)

data class ChatRequest(
    val model: String,
    val messages: List<ChatMessage>,
    val temperature: Double? = null,
    val maxTokens: Int? = null,
    val stream: Boolean = false
)

data class ChatChoice(
    val index: Int,
    val message: ChatMessage,
    val finishReason: String?
)

data class ChatResponse(
    val id: String,
    val model: String,
    val choices: List<ChatChoice>,
    val usage: TokenUsage?
)

data class TokenUsage(
    val promptTokens: Int,
    val completionTokens: Int,
    val totalTokens: Int
)

/**
 * Result returned to the caller after the full proxy pipeline.
 */
sealed class ProxyResult {
    /** Pipeline completed and LLM responded. */
    data class Success(val response: ChatResponse, val savings: ProxySavings? = null) : ProxyResult()

    /** A pipeline stage blocked the request (e.g. malicious prompt detected). */
    data class Blocked(val reason: BlockReason, val stage: String) : ProxyResult()

    /** An error occurred in the pipeline or upstream. */
    data class Error(val message: String, val cause: Throwable? = null) : ProxyResult()
}

enum class BlockReason {
    MALICIOUS_PROMPT,
    POLICY_VIOLATION,
    UPSTREAM_ERROR
}

// Pipeline-only result (no upstream call)

data class PipelineOnlyResult(
    val isBlocked: Boolean,
    val blockedBy: String?,
    val blockReason: String?,
    val processedMessages: List<ChatMessage>,
    val stageResults: List<StageResult>
)

data class StageResult(
    val name: String,
    val status: StageStatus,
    val detail: String? = null
)

enum class StageStatus {
    PASSED,
    MODIFIED,
    BLOCKED,
    ERROR
}

File: proxy-ios/Sources/MelangeLmProxy/Model/ChatModels.swift
//
//  ChatModels.swift
//  MelangeLmProxy
//
//  OpenAI-compatible chat request/response data models.
//

import Foundation

// MARK: - Request / Response

public struct ChatMessage: Codable, Equatable, Sendable {
    public let role: String     // "system" | "user" | "assistant"
    public let content: String

    public init(role: String, content: String) {
        self.role = role
        self.content = content
    }
}

public struct ChatRequest: Sendable {
    public let model: String
    public let messages: [ChatMessage]
    public let temperature: Double?
    public let maxTokens: Int?
    public let stream: Bool

    public init(
        model: String,
        messages: [ChatMessage],
        temperature: Double? = nil,
        maxTokens: Int? = nil,
        stream: Bool = false
    ) {
        self.model = model
        self.messages = messages
        self.temperature = temperature
        self.maxTokens = maxTokens
        self.stream = stream
    }
}

public struct ChatChoice: Sendable {
    public let index: Int
    public let message: ChatMessage
    public let finishReason: String?
}

public struct TokenUsage: Sendable {
    public let promptTokens: Int
    public let completionTokens: Int
    public let totalTokens: Int
}

public struct ChatResponse: Sendable {
    public let id: String
    public let model: String
    public let choices: [ChatChoice]
    public let usage: TokenUsage?
}

// MARK: - Proxy Result

public enum ProxyResult: Sendable {
    /// Pipeline completed and the LLM responded.
    case success(ChatResponse)
    /// A pipeline stage blocked the request before it reached the LLM.
    case blocked(reason: BlockReason, stage: String)
    /// An error occurred in the pipeline or upstream.
    case failure(message: String, error: Error?)
}

public enum BlockReason: Sendable {
    case maliciousPrompt
    case policyViolation
    case upstreamError
}

// MARK: - Pipeline-only result (no upstream call)

/// Result of running only the on-device pipeline stages, without calling the upstream LLM.
public struct PipelineOnlyResult: Sendable {
    public let isBlocked: Bool
    public let blockedBy: String?
    public let blockReason: String?
    /// Messages after all stages processed them (anonymized, summarized, etc.)
    public let processedMessages: [ChatMessage]
    /// Per-stage results showing what each stage did.
    public let stageResults: [StageResult]
}

public struct StageResult: Sendable {
    public let name: String
    public let status: StageStatus
    public let detail: String?

    public init(name: String, status: StageStatus, detail: String? = nil) {
        self.name = name
        self.status = status
        self.detail = detail
    }
}

public enum StageStatus: Sendable {
    case passed
    case modified
    case blocked(String)
    case error(String)
}
